### PR TITLE
fix(portal): keep the acting character across a page reload

### PR DESCRIPTION
**Pull request by HarryCordewener**
## The bug

`AccountAuthService.ActiveCharacter` lives only in memory. sessionStorage holds the token, username, role and permissions — nothing about who the tab is acting as — and `InitCoreAsync` never restored it. After a reload, `SetCharacters` re-seats the account on `characters.FirstOrDefault()`.

`ActingCharacterHeaderHandler` puts that value on the wire as `X-Acting-Character`, and the server routes per-character actions by it. So: switch to a non-primary character, press F5, and you are silently acting as your primary again. Gallery uploads, mail and wiki edits get attributed to a character the player wasn't acting as, with nothing on screen to say so.

Reproduced as a test before fixing — switch to Beta (#20), simulate a reload, read back:

```
AssertionException: Expected to be equal to 20 but received 10
```

## The fix

Mirror the active character into tab-scoped storage on every switch (`sharpmush.account.activeCharacter`) and read it back during hydration.

Two things worth calling out:

- **Restore only into an empty slot.** Nothing orders `InitAsync` against `LoginAsync`, so hydration can run *after* a login has already seated a character in memory. My first cut assigned unconditionally and broke 6 existing tests in `NewTabCharacterTests` / `GlobalTerminalIdentityTests` by clobbering a live active character with an empty read — a genuine flaw the suite caught, not a test artifact. Storage now fills an empty slot and never overwrites a live decision.
- **`SetCharacters` still re-validates against the roster**, so a restored character the account no longer owns is reseated exactly as before — covered by a test.

Persisting is deliberately not awaited. The setter is synchronous because all five callers depend on `ActiveCharacter` being visible the instant it returns, and storage is a cache of that decision rather than the decision itself. WASM is single-threaded and its interop calls run in issue order, so back-to-back switches land last-write-wins. Making it async would mean renaming through 5 production and 15 test call sites for no behavioural gain. A failed write is logged and costs only the next reload's acting character.

## Verification

- 3 new tests in `AccountAuthServiceActiveCharacterRefreshTests`, over a dictionary-backed `sessionStorage` fake so the store outlives the service that wrote to it — a per-call stubbed JSInterop can't express "survives a reload". Covers: survives refresh, dropped when the account no longer owns it, and nothing left behind after logout.
- Red/green confirmed by stashing the production change (verified the revert actually landed before trusting the run): `Expected 20 but received 10` without it, green with it.
- Full `SharpMUSH.Tests.BUnit` 273/273; `SharpMUSH.Tests` client namespace 111/111; `dotnet format whitespace` clean.

Follow-up to #715, same area — that one closed the bearer-token hydration window, this one closes the acting-character one.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

- **New Features**
  - The selected character remains active when refreshing the page within the same browser tab.
  - Previously active character details are automatically restored when available and valid.

- **Bug Fixes**
  - Restored selections are discarded if the account no longer owns that character.
  - Character changes during reload are no longer overwritten by stale stored values.
  - Logout clears the active character to prevent stale selections on later visits.
  - Storage read/write failures no longer disrupt character switching or initialization.

- **Tests**
  - Added automated coverage for persistence and hydration edge cases.

### CODE REVIEW

File: SharpMUSH.Client/Services/AccountAuthService.cs [hunk 1]
```diff
@@ -192,6 +243,18 @@ private async Task InitCoreAsync()
...
+		// Restore who this tab was acting as, but only into an empty slot: hydration can run AFTER a
+		// login has already seated a character in memory (nothing orders InitAsync against LoginAsync),
+		// and storage must never overwrite a live decision. Assigned directly rather than through
+		// SetActiveCharacter because the value came out of storage — writing it straight back is a
+		// pointless interop round-trip. SetCharacters re-validates it against the roster once that
+		// lands, so a character the account no longer owns is reseated like any other stale active.
+		if (ActiveCharacter is null && await ReadStoredActiveCharacterAsync() is { } restored)
+		{
+			ActiveCharacter = restored;
+			RaiseActiveCharacterChanged();
```
[the remark this answers is not in the document]
Comment (HarryCordewener) [reply SharpMUSH.Client/Services/AccountAuthService.cs]: Fixed in 00e186ec — you're right. Read into a local first, then re-check the slot:

```csharp
var restored = await ReadStoredActiveCharacterAsync();
if (restored is not null && ActiveCharacter is null)
```

Worth noting the test I wrote for it initially passed against the *broken* ordering, because it set the character before the read was actually in flight. The storage fake now signals when the gated read starts, and the test fails without this change (`Expected 20 but received 10`).

File: SharpMUSH.Client/Services/AccountAuthService.cs [hunk 2]
```diff
@@ -84,9 +85,59 @@ public void SetActiveCharacter(CharacterSummary? character)
...
+		try
+		{
+			return JsonSerializer.Deserialize<CharacterSummary>(stored);
+		}
+		catch (JsonException ex)
+		{
+			logger.LogWarning(ex, "Stored active character was unreadable; falling back to the roster default");
+			return null;
+		}
+	}
```
[the remark this answers is not in the document]
Comment (HarryCordewener) [reply SharpMUSH.Client/Services/AccountAuthService.cs:127]: Skipping this one. The guideline in CLAUDE.md is `Discriminated unions via OneOf<T1, T2> (never nullable returns from **services**)` — it governs service contracts, not private helpers. `ReadStoredActiveCharacterAsync` is a private method whose single caller immediately branches on the result, so a `OneOf<CharacterSummary, NoStoredActiveCharacter>` plus a marker type would add a type and a match for no callable surface and no extra safety.

File: SharpMUSH.Tests.BUnit/Services/AccountAuthServiceActiveCharacterRefreshTests.cs [hunk 3]
```diff
@@ -0,0 +1,154 @@
+using Microsoft.Extensions.Logging;
+using Microsoft.JSInterop;
+using Microsoft.JSInterop.Infrastructure;
+using NSubstitute;
```
[the remark this answers is not in the document]
Comment (HarryCordewener) [reply SharpMUSH.Tests.BUnit/Services/AccountAuthServiceActiveCharacterRefreshTests.cs:3]: Removed in 00e186ec — you're right, `IJSVoidResult` is only referenced through the `InvokeVoidAsync` extension, so the using was dead.

File: SharpMUSH.Client/Services/AccountAuthService.cs [hunk 4]
```diff
@@ -84,9 +85,59 @@ public void SetActiveCharacter(CharacterSummary? character)
...
+				await js.InvokeVoidAsync("sessionStorage.removeItem", ActiveCharacterKey);
+			else
+				await js.InvokeVoidAsync("sessionStorage.setItem", ActiveCharacterKey, JsonSerializer.Serialize(character));
+		}
+		catch (Exception ex)
+		{
+			// A failed write costs the player their acting character on the next reload — worth a log,
+			// never worth taking down the switch that already succeeded in memory.
+			logger.LogWarning(ex, "Could not persist the active character; a reload will fall back to the primary");
+		}
```
[the remark this answers is not in the document]
Comment (HarryCordewener) [reply SharpMUSH.Client/Services/AccountAuthService.cs:143]: Keeping the broad catch. This is a fire-and-forget cache write whose only job is to make the *next* reload accurate; nothing it can throw should reach the switch that already succeeded in memory, and during teardown JS interop throws `ObjectDisposedException`/`TaskCanceledException` as readily as `JSException` — enumerating those is a list that rots. There's direct precedent in the same file: `RaiseActiveCharacterChanged` catches `Exception` for the same reason (a subscriber's render exception must not propagate mid-switch). The failure is logged, not swallowed silently.

File: SharpMUSH.Tests.BUnit/Services/AccountAuthServiceActiveCharacterRefreshTests.cs [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+				accountId = "acct-1",
+				username = "headwiz",
+				characters,
+				accountSessionToken = "session-token-1",
+				mustChangePassword = false,
+				role = "God",
+				permissions = new[] { "*" },
+			});
+
+		return Task.FromResult(new HttpResponseMessage(HttpStatusCode.OK) { Content = content });
```
[the remark this answers is not in the document]
Comment (HarryCordewener) [reply SharpMUSH.Tests.BUnit/Services/AccountAuthServiceActiveCharacterRefreshTests.cs:109]: Skipping — same ownership false positive as on #715: the fake handler **returns** the `HttpResponseMessage` to its caller, which is what `HttpMessageHandler` requires. Disposing it before returning would hand `HttpClient` a dead response. 27 files in this project build and return a response from a fake handler the same way.

File: SharpMUSH.Tests.BUnit/Services/AccountAuthServiceActiveCharacterRefreshTests.cs [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+public class AccountAuthServiceActiveCharacterRefreshTests
+{
+	private static readonly CharacterSummary Alpha = new(10, 1000, "Alpha", "PLAYER");
+	private static readonly CharacterSummary Beta = new(20, 2000, "Beta", "PLAYER");
+
+	private static AccountAuthService MakeService(IJSRuntime js, IReadOnlyList<CharacterSummary> roster)
+	{
+		var httpClientFactory = Substitute.For<IHttpClientFactory>();
+		// Not disposed: the returned service keeps calling through this client after the helper returns.
+		var http = new HttpClient(new FakeAccountApiHandler(roster)) { BaseAddress = new Uri("https://localhost:8081/") };
```
[the remark this answers is not in the document]
Comment (HarryCordewener) [reply SharpMUSH.Tests.BUnit/Services/AccountAuthServiceActiveCharacterRefreshTests.cs:129]: Skipping. The `HttpClient` is deliberately kept alive — the service returned by `MakeService` keeps calling through it after the helper returns, which is why the line carries a comment saying so. It's the established pattern in this project's auth tests (`AccountAuthServiceActiveCharacterTests.MakeUnlinkableServiceAsync` does the same with the same note). Tracking disposables for teardown would add fixture plumbing to short-lived tests whose handlers hold no sockets.

File: SharpMUSH.Client/Services/AccountAuthService.cs [hunk 7]
```diff
@@ -192,6 +243,22 @@ private async Task InitCoreAsync()
...
+		// lands, so a character the account no longer owns is reseated like any other stale active.
+		// Re-check the slot AFTER the read: the storage round-trip is an await, and a login can seat a
+		// character while it is in flight. Testing before the await would let a stale stored value
+		// clobber the newer live one.
+		var restored = await ReadStoredActiveCharacterAsync();
+		if (restored is not null && ActiveCharacter is null)
+		{
+			ActiveCharacter = restored;
+			RaiseActiveCharacterChanged();
+		}
```
[the remark this answers is not in the document]
Comment (HarryCordewener) [reply SharpMUSH.Client/Services/AccountAuthService.cs]: Applied in 6a12fa2 — and you're right that it composes with the post-await re-check rather than replacing it. The read is now guarded by an up-front `ActiveCharacter is null`, with the re-check kept inside it.

Both are load-bearing: I deleted each in turn to check. Without the up-front guard we spend an interop call to learn something already known; without the re-check, `ALoginLandingDuringHydration_IsNotClobberedByTheStoredValue` fails.

File: SharpMUSH.Client/Services/AccountAuthService.cs [hunk 8]
```diff
@@ -84,9 +85,59 @@ public void SetActiveCharacter(CharacterSummary? character)
...
+
+		try
+		{
+			return JsonSerializer.Deserialize<CharacterSummary>(stored);
+		}
+		catch (JsonException ex)
+		{
+			logger.LogWarning(ex, "Stored active character was unreadable; falling back to the roster default");
+			return null;
+		}
```
[the remark this answers is not in the document]
Comment (HarryCordewener) [reply SharpMUSH.Client/Services/AccountAuthService.cs:126]: Applied in 6a12fa2 — the key is cleared on `JsonException` now, so it self-heals instead of re-logging on every reload.

Covered by a new test, though my first assertion for it was wrong: I asserted the key was simply absent afterwards, but the fallback immediately re-seats storage with the roster default, so it legitimately comes back holding a valid value. The test now checks both halves — dropped at parse failure, then re-seated with Alpha.

File: SharpMUSH.Client/Services/AccountAuthService.cs [hunk 9]
```diff
@@ -84,9 +85,62 @@ public void SetActiveCharacter(CharacterSummary? character)
 			return;
 
 		ActiveCharacter = character;
+		PersistActiveCharacter(character);
 		RaiseActiveCharacterChanged();
```
[the remark this answers is not in the document]
Comment (HarryCordewener) [reply SharpMUSH.Client/Services/AccountAuthService.cs:89]: Fixed in a745518 — you're right that this broke the invariant that comment states. `LogoutAsync` now awaits `removeItem(ActiveCharacterKey)` alongside the other keys, so the removal is complete before `RaiseAuthStateChanged` fires.

Being straight about the test: it covers the end state (no stored character survives a logout), not the ordering. In-process the detached write also lands before `LogoutAsync` returns, so the test passes with or without the awaited removal — I checked, and its comment now says so rather than overclaiming. The awaited removal is what makes the invariant hold in a browser, where a navigation right after logout can cut a detached continuation short.

File: SharpMUSH.Client/Services/AccountAuthService.cs [hunk 10]
```diff
@@ -192,6 +246,27 @@ private async Task InitCoreAsync()
...
+		// then re-check the slot AFTER the read as well: the storage round-trip is an await, and a login
+		// can land while it is in flight. Testing only before the await would let a stale stored value
+		// clobber the newer live one; testing only after it would spend an interop call to learn
+		// something we already knew.
+		if (ActiveCharacter is null)
+		{
+			var restored = await ReadStoredActiveCharacterAsync();
+			if (restored is not null && ActiveCharacter is null)
+			{
+				ActiveCharacter = restored;
```
[the remark this answers is not in the document]
Comment (HarryCordewener) [reply SharpMUSH.Client/Services/AccountAuthService.cs:289]: Fixed in a745518, taking your second option. `SetCharacters` now re-binds a still-present active character to the roster's instance (records compare by value, so a drifted name or flag set triggers it) and re-persists, so the next reload restores the fresh copy too.

I'd considered rebinding when I first wrote `SetCharacters` and skipped it as cosmetic; you're right that persistence changes the calculus, since a restored snapshot can now predate a rename by any amount of time. Covered by a test that renames the character between the switch and the reload — without the rebind it reports "Beta" where the roster says "BetaRenamed".

File: SharpMUSH.Client/Services/AccountAuthService.cs [hunk 11]
```diff
@@ -640,6 +734,15 @@ public async Task LogoutAsync()
...
 		await js.InvokeVoidAsync("sessionStorage.removeItem", PermissionsKey);
+		// Awaited here rather than left to SetActiveCharacter's detached write above: the invariant
+		// below (every storage mutation complete before the event fires) has to hold for this key too,
+		// or a stale acting character could outlive the logout and be restored by the next login in
+		// this tab.
+		await js.InvokeVoidAsync("sessionStorage.removeItem", ActiveCharacterKey);
+		// Awaited here rather than left to SetActiveCharacter's detached write above: the invariant
+		// below (every storage mutation complete before the event fires) has to hold for this key too,
+		// or a stale acting character could outlive the logout and be restored by the next login in
+		// this tab.
```
[the remark this answers is not in the document]
Comment (HarryCordewener) [reply SharpMUSH.Client/Services/AccountAuthService.cs:746]: Removed — thanks. It was debris from a botched revert-and-restore while I was red-checking the previous commit: the comment got re-inserted while the original block's `await` line had already been stripped, leaving the orphan.

I also grepped the rest of the diff for duplicated added lines in case that revert left anything else behind; nothing but braces and `try`.

File: SharpMUSH.Client/Services/AccountAuthService.cs [hunk 12]
```diff
@@ -192,6 +265,27 @@ private async Task InitCoreAsync()
...
+		// lands, so a character the account no longer owns is reseated like any other stale active.
+		// Skip the read outright when a character is already seated (a login can beat hydration here),
+		// then re-check the slot AFTER the read as well: the storage round-trip is an await, and a login
+		// can land while it is in flight. Testing only before the await would let a stale stored value
+		// clobber the newer live one; testing only after it would spend an interop call to learn
+		// something we already knew.
+		if (ActiveCharacter is null)
+		{
+			var restored = await ReadStoredActiveCharacterAsync();
+			if (restored is not null && ActiveCharacter is null)
```
[the remark this answers is not in the document]
Comment (HarryCordewener) [reply SharpMUSH.Client/Services/AccountAuthService.cs:287]: Not removing this one — the condition is only constant if you don't model the `await` between the two checks.

`ReadStoredActiveCharacterAsync()` is a JS interop round-trip. Another continuation (a login completing) can run during it and seat a character, which is precisely what the re-check catches. It's verified, not assumed: deleting the inner `ActiveCharacter is null` makes `ALoginLandingDuringHydration_IsNotClobberedByTheStoredValue` fail with `Expected 20 but received 10`. That test holds the stored-character read open until a switch has landed.

The outer check is the one that saves the interop call; the inner one saves the selection. I've added a comment naming the test so this doesn't get tidied away later (1454b25a).

File: SharpMUSH.Client/Services/AccountAuthService.cs [hunk 13]
```diff
@@ -84,9 +85,62 @@ public void SetActiveCharacter(CharacterSummary? character)
...
+		try
+		{
+			return JsonSerializer.Deserialize<CharacterSummary>(stored);
+		}
+		catch (JsonException ex)
+		{
+			// Drop the bad entry so this self-heals: left in place it would re-log on every reload
+			// until the player happened to switch characters.
+			logger.LogWarning(ex, "Stored active character was unreadable; clearing it and falling back to the roster default");
+			PersistActiveCharacter(null);
```
[the remark this answers is not in the document]
Comment (HarryCordewener) [reply SharpMUSH.Client/Services/AccountAuthService.cs]: Fixed in bbf9492 — `await PersistActiveCharacterCoreAsync(null)`. You're right that the method is already async, so there's no reason for the recovery clear to be detached, and a navigation cutting it short would undo the self-heal it exists for.

Chasing this also exposed that my test wasn't proving anything. It asserted the end state, and the detached write happened to land before the assertion anyway — so it passed against both versions. The storage fake can now park writes, and the test asserts hydration *cannot finish* until the clear it issued completes; it fails against the detached version.

Same lesson as the logout clear you flagged earlier: where the code is already async, await the write and let the test prove the ordering.

File: SharpMUSH.Tests.BUnit/Services/AccountAuthServiceActiveCharacterRefreshTests.cs [hunk 14]
```diff
@@ -0,0 +1,296 @@
...
+		var seed = MakeService(storage, roster);
+		await seed.InitAsync();
+		await seed.LoginAsync("headwiz", "password-one");
+		seed.SetActiveCharacter(Alpha);
+
+		// Reload. Hold the stored-character read open so a login can land mid-hydration.
+		var afterRefresh = MakeService(storage, roster);
+		storage.GatedKey = "sharpmush.account.activeCharacter";
+		storage.GateReadsOf = new TaskCompletionSource();
+		storage.ReadStarted = new TaskCompletionSource();
```
[the remark this answers is not in the document]
Comment (HarryCordewener) [reply SharpMUSH.Tests.BUnit/Services/AccountAuthServiceActiveCharacterRefreshTests.cs]: Applied to all three gates.

One correction to the premise: the convention here is actually mixed — three plain `new TaskCompletionSource()` against three with the flag. So this isn't a house style I'd drifted from.

It's still the right call on the merits, and for one of the three it's more than style: `ReadStarted` is completed from *inside* the fake's `InvokeAsync`, so an inline continuation would resume the test body reentrantly in the middle of the interop call it is signalling about. Ran the gated tests three times after the change to be sure the scheduling shift didn't introduce flakiness.
